esp32/capstone/src/main.cpp

```diff
@@ -0,0 +1,283 @@
+
+
+/* Hall effect sensor interrupt service routine */
+void IRAM_ATTR HallEffectSensor_ISR()
+{
+    portENTER_CRITICAL_ISR(&hesMux);
+    g_newRotation = true;
+    g_revolutions++;
+    portEXIT_CRITICAL_ISR(&hesMux);
+}
+
+/* ISR for HW timer */
+void IRAM_ATTR Timer90Deg_ISR()
+{
+    portENTER_CRITICAL_ISR(&timerMux);
+    for (int i = 0; i < 4; i++)
+    {
+        if (g_arm_quarter_map[i] == 3)
+        {            
+            g_arm_quarter_map[i] =0;
+        }
+        else
+        {
+            g_arm_quarter_map[i]++;
+        } 
+    }
+    portEXIT_CRITICAL_ISR(&timerMux);
+}
+
+
+void CalculateRPM()
+{
+    // do new calculation every 3 revolutions - decrease for improved accuracy
+    if (g_revolutions >= 3)
+    {
+        if (g_prevTimeMs != 0) // dont divide by 0
+            g_revsPerMs = g_revolutions/((float)(millis() - g_prevTimeMs));
+        g_prevTimeMs = millis();
+        g_revolutions = 0;
+
+        g_currentRpm = g_revsPerMs * 60000;
+        if (g_revsPerMs != 0) // dont divide by zero
+        {
+            g_timePer90degMs = (1/g_revsPerMs) / 4;
+            g_timePerSegment_us = ((1/g_revsPerMs) / NUM_ANGULAR_SEGMENTS)*1000;
+        }
+        DEBUG_PRINT("Calculations------");
+        DEBUG_PRINT(g_currentRpm);
+        DEBUG_PRINT(g_timePer90degMs);
+        DEBUG_PRINT(g_timePerSegment_us);
+        DEBUG_PRINT("------------------");
+    }
+}
+
+
+void Set90degTimerInterval(int interval_us)
+{
+    timerAlarmWrite(quarter_circle_timer, interval_us, true);
+    timerAlarmEnable(quarter_circle_timer);
+}
+
+
+void Task_ImageDisplay(void *parameter)
+{
+    DEBUG_PRINT("Start Image Display Task");
+
+    ImageFrameData *imageFrame;
+#if DEMO
+        imageFrame = demoImage;
+#endif
+
+    for (;;) // infinite loop
+    {
+        
+#if DEMO == 0
+        xQueueReceive(ImageFrameQueue, imageFrame, 4);
+#endif
+
+        CalculateRPM(); // update params
+        g_newRotation = false;
+
+        if (g_currentRpm >= (float)MOTOR_RPM_THRESHOLD)
+            DEBUG_PRINT("***************** NEW ROTATION *****************");
+        
+        while ((!g_newRotation) && (g_currentRpm >= (float)MOTOR_RPM_THRESHOLD)) // loop for 1 rotation
+        {
+            DEBUG_PRINT(g_arm_quarter_map[0]);
+            
+            Set90degTimerInterval(g_timePer90degMs*1000);
+
+            // display image on LEDS
+            for (int segmentIdx = 0; segmentIdx < NUM_ANGULAR_SEGMENTS; segmentIdx++)
+            {
+                for (int i = 0; i < 4; i++) // iterate over arms
+                {
+                    for (int j = 0; j < NUM_LEDS_ARM; j++) // iterate over leds within an arm
+                    {
+                        int ledIdx = i*NUM_LEDS_ARM + j;
+                        int segmentOffset = g_arm_quarter_map[i]*NUM_LEDS_ARM;
+                        CRGB rgbData = imageFrame->ledValues[i][segmentOffset + segmentIdx];
+                        g_leds[ledIdx] = rgbData;
+                    }
```

Comment:
per the PR description, I saw that you hadn't yet gotten a chance to do testing here, but I noticed the  couple of the following thing:

It looks like this inner loop just copies `imageFrame->ledValues[i][segmentOffset + segmentIdx]` into `g_leds[i*NUM_LEDS_ARM...i*NUM_LEDS_ARM+NUM_LEDS_ARM]`. I believe that you meant to do copy `imageFrame->ledValues[j][segmentOffset + segmentIdx]` based on data structure sizes. I have no idea if this is just a bug fix vs. a wonky observation that won't be relevant since this code is untested, but it _seems_ fishy in the context of the current code.